### PR TITLE
Suport typeFilter parameter for bulk export

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkExportGenrateResourceFilesStepListener.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import org.springframework.batch.core.ExitStatus;
+import org.springframework.batch.core.StepExecution;
+import org.springframework.batch.core.StepExecutionListener;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import javax.annotation.Nonnull;
+
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+public class BulkExportGenrateResourceFilesStepListener  implements StepExecutionListener {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkExportGenrateResourceFilesStepListener.java:14]: Genrate -> Generate

Comment (tadgh) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkExportGenrateResourceFilesStepListener.java:14]: Maybe add a small class javadoc indicating that this class is meant to set correct failure state on the entity. It's obvious enough reading the afterStep, but a small docstring could help

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkItemReader.java [hunk 2]
```diff
@@ -31,73 +32,78 @@
...
+		Map<String, String[]> requestUrl = UrlUtil.parseQueryStrings(jobEntity.getRequest());
+		String[] typeFilters = requestUrl.get(JpaConstants.PARAM_EXPORT_TYPE_FILTER);
+		if (typeFilters != null) {
+			Optional<String> filter = Arrays.stream(typeFilters).filter(t -> t.startsWith(myResourceType + "?")).findFirst();
+			if (filter.isPresent()) {
+				String matchUrl = filter.get();
+				map = myMatchUrlService.translateMatchUrl(matchUrl, def);
+			}
+		}
+
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkItemReader.java:101]: This logical block could be ripped out into a helper function for conciseness.

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java [hunk 3]
```diff
@@ -290,14 +284,35 @@ public JobInfo submitJob(String theOutputFormat, Set<String> theResourceTypes, D
...
+				}
+				String resourceType = next.substring(0, next.indexOf("?"));
+				if (!resourceTypes.contains(resourceType)) {
+					throw new InvalidRequestException("Invalid " + JpaConstants.PARAM_EXPORT_TYPE_FILTER + " value \"" + next + "\". Resource type does not appear in " + JpaConstants.PARAM_EXPORT_TYPE + " list");
+				}
+				if (!types.add(resourceType)) {
+					throw new InvalidRequestException("Invalid " + JpaConstants.PARAM_EXPORT_TYPE_FILTER + " value \"" + next + "\". Multiple filters found for type " + resourceType);
+				}
+			}
+		}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java:310]: Another good candidate for helper extraction

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java [hunk 4]
```diff
@@ -374,4 +388,14 @@ public synchronized void cancelAndPurgeAllJobs() {
...
+
+	public static class Job implements HapiJob {
+		@Autowired
+		private IBulkDataExportSvc myTarget;
+
+		@Override
+		public void execute(JobExecutionContext theContext) {
+			myTarget.buildExportFiles();
+		}
+	}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java:400]: Do we still need this? Trying to recall where this is used

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkExportDaoSvc.java [hunk 5]
```diff
@@ -120,96 +104,9 @@ public void setJobToStatus(String theJobUUID, BulkJobStatusEnum theStatus) {
...
-		return toSubmittedJobInfo(job);
-	}
-
-	private void updateExpiry(BulkExportJobEntity theJob) {
-		theJob.setExpiry(DateUtils.addMilliseconds(new Date(), myRetentionPeriod));
-	}
-	private IBulkDataExportSvc.JobInfo toSubmittedJobInfo(BulkExportJobEntity theJob) {
-		return new IBulkDataExportSvc.JobInfo().setJobId(theJob.getJobId());
-	}
-
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkExportDaoSvc.java:214]: This was all just the old method right?

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java [hunk 6]
```diff
@@ -156,7 +162,89 @@ public void testSubmit_InvalidResourceTypes() {
...
+			myBulkDataExportSvc.buildExportFiles();
+
+			awaitAllBulkJobCompletions();
+
+			// Fetch the job again
+			status = myBulkDataExportSvc.getJobInfoOrThrowResourceNotFound(jobDetails.getJobId());
+			assertEquals(BulkJobStatusEnum.ERROR, status.getStatus());
+			assertThat(status.getStatusMessage(), containsString("help i'm a bug"));
+
+		} finally {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java:240]: This is an awesome test

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java [hunk 7]
```diff
@@ -400,15 +437,95 @@ public void testSubmit_WithSince() throws InterruptedException {
...
+			myBatchJobSubmitter.runJob(myBulkJob, paramBuilder.toJobParameters());
+			fail("Should have had invalid parameter execption!");
+		} catch (JobParametersInvalidException e) {
+			// good
+		}
+
+	}
+	//Note that if the job is generated, and doesnt rely on an existed persisted BulkExportJobEntity, it will need to
+	//create one itself, which means that its jobUUID isnt known until it starts. to get around this, we move
+
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java:508]: Did I just never complete this comment?!

**Review by tadgh — APPROVED** [review]
Looks good, couple questions and minor refactors. Approved! :tada:

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkExportGenrateResourceFilesStepListener.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkExportGenrateResourceFilesStepListener.java:14]: Done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkItemReader.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/job/BulkItemReader.java:101]: done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java:310]: done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java (hunk 4 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkDataExportSvcImpl.java:400]: I believe it's needed.. this is the scheduled job that looks for job entities that aren't in the job executor yet and requests the batch job to start

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkExportDaoSvc.java (hunk 5 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/bulk/svc/BulkExportDaoSvc.java:214]: yeah

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java (hunk 6 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java:240]: image

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java (hunk 7 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/bulk/BulkDataExportSvcImplR4Test.java:508]: heh looks like. i'll complete it

**Pull request merged**